tick/optim/solver/src/svrg.cpp

```diff
@@ -68,19 +70,24 @@ void SVRG::compute_step_corrections() {
 }
 
 void SVRG::solve_dense() {
-  for (ulong t = 0; t < epoch_size; ++t) {
-    ulong i = get_next_i();
-    model->grad_i(i, iterate, grad_i);
-    model->grad_i(i, fixed_w, grad_i_fixed_w);
-    for (ulong j = 0; j < iterate.size(); ++j) {
-      iterate[j] = iterate[j] - step * (grad_i[j] - grad_i_fixed_w[j] + full_gradient[j]);
-    }
-    prox->call(iterate, step, iterate);
-    if (variance_reduction == VarianceReductionMethod::Random && t == rand_index) {
-      next_iterate = iterate;
-    }
-    if (variance_reduction == VarianceReductionMethod::Average) {
-      next_iterate.mult_incr(iterate, 1.0 / epoch_size);
+  if (threads > 1) {
```

Comment:
Just wondering, don't you think we could isolate this code in a function and pass `dense_single_thread_solver` / `sparse_single_thread_solver` as argument? It would avoid code duplication and could afterwards probably be shared with all sto solvers

Reply:
can we put that in a different ticket - I prefer simplicity but that the method parameters are not the same it introduces difficulty

Reply:
Makes sense, it will mostly be useful once we generalize asynchronous solving to all stochastic solvers.